### ISSUE TITLE
ImportWarning: sys.meta_path is empty

### ISSUE THREAD

**Issue by foxx** [ipython/ipython#8828]
In some situations, if you run unittests after importing IPython, it will cause the following errors to be flooded in `3.4`;

```
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='/dev/null' mode='w' encoding='UTF-8'>
/home/vagrant/.virtualenvs/vagrant/lib/python3.4/importlib/_bootstrap.py:2127: ImportWarning: sys.meta_path is empty
/home/vagrant/.virtualenvs/vagrant/lib/python3.4/importlib/_bootstrap.py:2127: ImportWarning: sys.meta_path is empty
/home/vagrant/.virtualenvs/vagrant/lib/python3.4/importlib/_bootstrap.py:2127: ImportWarning: sys.meta_path is empty
/home/vagrant/.virtualenvs/vagrant/lib/python3.4/importlib/_bootstrap.py:2127: ImportWarning: sys.meta_path is empty
... repeated hundreds of times ...
```

There is a fixed [bug](https://bugs.python.org/issue21049) in Python and another repo with the same [problem](https://github.com/vispy/vispy/issues/555).

If I remove the following import line, the problem goes away;

```
from IPython import start_ipython
```

Clearly there is something not quite right with the way IPython has been packaged, importing a function from a package should not result in any runtime execution differences.

**Comment by minrk** [comment]
Can you provide a complete example that reproduces the issue? The only interaction IPython has with sys.meta_path is adding import hooks when deprecated modules are imported. It never removes items from sys.meta_path.

**Comment by jakirkham** [comment]
The `sys.meta_path` is an upstream issue in Python. It has been corrected in Python 3.5. I am not sure if a bugfix patch will be added for Python 3.4 or not. See more details here ( https://bugs.python.org/issue21049 ).

The `ResourceWarning` is caused by this line ( https://github.com/ipython/ipython/blob/a33a0632c00b604683aab8454983716823550407/IPython/utils/io.py#L75 ) and can be fixed with this PR ( https://github.com/ipython/ipython/pull/8929 ).

**Comment by foxx** [comment]
Nice work, thanks guys!
